In [3]:
import pandas as pd
import numpy as np

In [47]:
df = pd.read_csv('train_features.csv')
df.head().T

0                        1  \
id                                       69572                     8776   
amount_tsh                                6000                        0   
date_recorded                       2011-03-14               2013-03-06   
funder                                   Roman                  Grumeti   
gps_height                                1390                     1399   
installer                                Roman                  GRUMETI   
longitude                              34.9381                  34.6988   
latitude                              -9.85632                 -2.14747   
wpt_name                                  none                 Zahanati   
num_private                                  0                        0   
basin                               Lake Nyasa            Lake Victoria   
subvillage                            Mnyusi B                  Nyamara   
region                                  Iringa                     Mara   
region_code                                 11                       20   
district_code                                5                        2   
lga                                     Ludewa                Serengeti   
ward                                  Mundindi                    Natta   
population                                 109                      280   
public_meeting                            True                      NaN   
recorded_by            GeoData Consultants Ltd  GeoData Consultants Ltd   
scheme_management                          VWC                    Other   
scheme_name                              Roman                      NaN   
permit                                   False                     True   
construction_year                         1999                     2010   
extraction_type                        gravity                  gravity   
extraction_type_group                  gravity                  gravity   
extraction_type_class                  gravity                  gravity   
management                                 vwc                      wug   
management_group                    user-group               user-group   
payment                           pay annually                never pay   
payment_type                          annually                never pay   
water_quality                             soft                     soft   
quality_group                             good                     good   
quantity                                enough             insufficient   
quantity_group                          enough             insufficient   
source                                  spring     rainwater harvesting   
source_type                             spring     rainwater harvesting   
source_class                       groundwater                  surface   
waterpoint_type             communal standpipe       communal standpipe   
waterpoint_type_group       communal standpipe       communal standpipe   

                                                 2  \
id                                           34310   
amount_tsh                                      25   
date_recorded                           2013-02-25   
funder                                Lottery Club   
gps_height                                     686   
installer                             World vision   
longitude                                  37.4607   
latitude                                  -3.82133   
wpt_name                               Kwa Mahundi   
num_private                                      0   
basin                                      Pangani   
subvillage                                 Majengo   
region                                     Manyara   
region_code                                     21   
district_code                                    4   
lga                                      Simanjiro   
ward                                       Ngorika 

In [48]:
df.shape

(59400, 40)

In [49]:
df_labels = pd.read_csv('train_labels.csv')
df_labels.head()

id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

In [50]:
df_labels.shape

(59400, 2)

In [12]:
#df_merged = pd.merge(df, df_labels, on='id')
#df_merged.head().T

0                        1  \
id                                       69572                     8776   
amount_tsh                                6000                        0   
date_recorded                       2011-03-14               2013-03-06   
funder                                   Roman                  Grumeti   
gps_height                                1390                     1399   
installer                                Roman                  GRUMETI   
longitude                              34.9381                  34.6988   
latitude                              -9.85632                 -2.14747   
wpt_name                                  none                 Zahanati   
num_private                                  0                        0   
basin                               Lake Nyasa            Lake Victoria   
subvillage                            Mnyusi B                  Nyamara   
region                                  Iringa                     Mara   
region_code                                 11                       20   
district_code                                5                        2   
lga                                     Ludewa                Serengeti   
ward                                  Mundindi                    Natta   
population                                 109                      280   
public_meeting                            True                      NaN   
recorded_by            GeoData Consultants Ltd  GeoData Consultants Ltd   
scheme_management                          VWC                    Other   
scheme_name                              Roman                      NaN   
permit                                   False                     True   
construction_year                         1999                     2010   
extraction_type                        gravity                  gravity   
extraction_type_group                  gravity                  gravity   
extraction_type_class                  gravity                  gravity   
management                                 vwc                      wug   
management_group                    user-group               user-group   
payment                           pay annually                never pay   
payment_type                          annually                never pay   
water_quality                             soft                     soft   
quality_group                             good                     good   
quantity                                enough             insufficient   
quantity_group                          enough             insufficient   
source                                  spring     rainwater harvesting   
source_type                             spring     rainwater harvesting   
source_class                       groundwater                  surface   
waterpoint_type             communal standpipe       communal standpipe   
waterpoint_type_group       communal standpipe       communal standpipe   
status_group                        functional               functional   

                                                 2  \
id                                           34310   
amount_tsh                                      25   
date_recorded                           2013-02-25   
funder                                Lottery Club   
gps_height                                     686   
installer                             World vision   
longitude                                  37.4607   
latitude                                  -3.82133   
wpt_name                               Kwa Mahundi   
num_private                                      0   
basin                                      Pangani   
subvillage                                 Majengo   
region                                     Manyara   
region_code                                     21   
district_code                                    4   
lga                           

In [53]:
df_labels.status_group.value_counts(normalize=True)

functional                 0.543081
non functional             0.384242
functional needs repair    0.072677
Name: status_group, dtype: float64

So we would expect the following percentages for these outcomes:
functional: 54.3%
non functional  38.4%
functional needs repair 7.3%

In [21]:
#df_merged.dtypes

id                         int64
amount_tsh               float64
date_recorded             object
funder                    object
gps_height                 int64
installer                 object
longitude                float64
latitude                 float64
wpt_name                  object
num_private                int64
basin                     object
subvillage                object
region                    object
region_code                int64
district_code              int64
lga                       object
ward                      object
population                 int64
public_meeting            object
recorded_by               object
scheme_management         object
scheme_name               object
permit                    object
construction_year          int64
extraction_type           object
extraction_type_group     object
extraction_type_class     object
management                object
management_group          object
payment                   object
payment_ty

In [54]:
df_num = df.select_dtypes(include=['number'])
print(df_num.shape)
df_num.head()

(59400, 10)


id  amount_tsh  gps_height  longitude   latitude  num_private  \
0  69572      6000.0        1390  34.938093  -9.856322            0   
1   8776         0.0        1399  34.698766  -2.147466            0   
2  34310        25.0         686  37.460664  -3.821329            0   
3  67743         0.0         263  38.486161 -11.155298            0   
4  19728         0.0           0  31.130847  -1.825359            0   

   region_code  district_code  population  construction_year  
0           11              5         109               1999  
1           20              2         280               2010  
2           21              4         250               2009  
3           90             63          58               1986  
4           18              1           0                  0

In [55]:
df_num.isnull().sum()

id                   0
amount_tsh           0
gps_height           0
longitude            0
latitude             0
num_private          0
region_code          0
district_code        0
population           0
construction_year    0
dtype: int64

In [46]:
# *** dont do this***
# 1hot encode status_group column
one_hot_status = pd.get_dummies(df_merged.status_group)
# drop the sex column
df_merged = df_merged.drop('status_group',axis = 1)
# join the get_dummies columns to the df
df_merged = df_merged.join(one_hot_status)
df_merged.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ...      quantity  \
0  34.938093  -9.856322                  none            0  ...        enough   
1  34.698766  -2.147466              Zahanati            0  ...  insufficient   
2  37.460664  -3.821329           Kwa Mahundi            0  ...        enough   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...           dry   
4  31.130847  -1.825359               Shuleni            0  ...      seasonal   

  quantity_group                source           source_type  source_class  \
0         enough                spring                spring   groundwater   
1   insufficient  rainwater harvesting  rainwater harvesting       surface   
2         enough                   dam                   dam       surface   
3            dry           machine dbh              borehole   groundwater   
4       seasonal  rainwater harvesting  rainwater harvesting       surface   

               waterpoint_type waterpoint_type_group  functional  \
0           communal standpipe    communal standpipe           1   
1           communal standpipe    communal standpipe           1   
2  communal standpipe multiple    communal standpipe           1   
3  communal standpipe multiple    communal standpipe           0   
4           communal standpipe    communal standpipe           1   

  functional needs repair non functional  
0                       0              0  
1                       0              0  
2                       0              0  
3                       0              1  
4                       0              0  

[5 rows x 43 columns]

In [57]:
X = df_num

y = df_labels.status_group

In [58]:
from sklearn.feature_selection import f_regression, SelectKBest, f_classif
from sklearn.linear_model import Ridge, LogisticRegression, LogisticRegressionCV
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import accuracy_score, classification_report

In [64]:
model = LogisticRegression(solver='lbfgs')
model.fit(X, y)
y_pred = model.predict(X)
print(classification_report(y, y_pred))
print('accuracy', accuracy_score(y, y_pred))

C:\Users\Brian\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
C:\Users\Brian\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
C:\Users\Brian\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


                         precision    recall  f1-score   support

             functional       0.56      0.93      0.70     32259
functional needs repair       0.00      0.00      0.00      4317
         non functional       0.49      0.12      0.19     22824

              micro avg       0.55      0.55      0.55     59400
              macro avg       0.35      0.35      0.30     59400
           weighted avg       0.49      0.55      0.45     59400

accuracy 0.5518013468013468


In [66]:
prediction = y_pred

array(['functional', 'functional', 'functional', ..., 'functional',
       'functional', 'functional'], dtype=object)

In [80]:
pred = pd.DataFrame(prediction)
pred.head()

0
0      functional
1      functional
2      functional
3  non functional
4      functional

In [81]:
pred = pd.concat([df_num.id, pred], axis=1)
pred.head()

id               0
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

In [83]:
pred.columns.values[1]= 'status_group'
pred.head()

id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

In [89]:
pred.describe

<bound method NDFrame.describe of           id    status_group
0      69572      functional
1       8776      functional
2      34310      functional
3      67743  non functional
4      19728      functional
5       9944      functional
6      19816      functional
7      54551      functional
8      53934      functional
9      46144      functional
10     49056  non functional
11     50409      functional
12     36957      functional
13     50495      functional
14     53752      functional
15     61848      functional
16     48451      functional
17     58155      functional
18     34169      functional
19     18274      functional
20     48375      functional
21      6091  non functional
22     58500      functional
23     37862      functional
24     51058      functional
25     22308      functional
26     55012  non functional
27     20145      functional
28     19685      functional
29     69124      functional
...      ...             ...
59370  14796      functional
59371  20

In [86]:
pred.to_csv('pred.csv', sep=',', encoding='utf-8', index=False)

End of initial submission. Onward to feature engineering and stuffs!

In [30]:
X_train = df_merged.drop(columns='status_group')
y_train = df_merged.status_group

X_test  = df_merged.drop(columns='status_group')
y_test  = df_merged.status_group

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=15)

In [37]:
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((44550, 10), (44550,), (14850, 10), (14850,))

In [42]:
y_train.head()

42161    non functional
12164        functional
54879    non functional
9059         functional
47419        functional
Name: status_group, dtype: object

In [39]:
# Define an estimator and param_grid
pipe = make_pipeline(
    RobustScaler(), 
    SelectKBest(f_classif), 
    LogisticRegression(solver='lbfgs'))

In [41]:
param_grid = {
    'selectkbest__k': range(1, len(X_train.columns)),
    'logisticregression__C': [.0001, .001, .01, .1, 1.0, 10.0, 100.00, 1000.0, 10000.0],
    'logisticregression__class_weight': [None, 'balanced']
}

# Fit on the train set, with grid search cross-validation
gs = GridSearchCV(pipe, param_grid=param_grid, cv=5, 
                  scoring='neg_mean_absolute_error', 
                  verbose=1)

gs.fit(X_train, y_train)

Fitting 5 folds for each of 162 candidates, totalling 810 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\Brian\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


ValueError: could not convert string to float: 'non functional'

(14850,)

In [ ]:
coefficients = model.coef_
feature_names = X_validate_ions.columns
pd.Series(coefficients, feature_names).sort_values()